File: 0_100/084_Largest_Rectangle_in_Histogram/Solution.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.insert(0,0)
        heights.append(0)
        n, ans, st = len(heights), 0, []
        for i in range(0,n):
            while len(st) != 0 and heights[i] < heights[st[len(st) - 1]]:
                h = heights[st.pop()]
                ans = max(ans, (i - st[len(st) - 1] -1) * h)
            st.append(i)
        return ans

File: 0_100/084_Largest_Rectangle_in_Histogram/Solution.rs
impl Solution {
    pub fn largest_rectangle_area(mut heights: Vec<i32>) -> i32 {
        heights.insert(0, 0);
        heights.push(0);
        let (n, mut ans) = (heights.len(), 0_i32);
        let mut st = vec![];
        for i in 0..n {
            while !st.is_empty() && heights[st[st.len() - 1]] > heights[i] {
                let h = heights[st.pop().unwrap()];
                ans = ans.max((i - st[st.len() - 1] - 1) as i32 * h);
            }
            st.push(i);
        }
        ans
    }
}